### DISCUSSION TITLE
Evaluating Superpower vs Sprache for my needs

### DISCUSSION THREAD

**Discussion by RebelusQuo** [datalust/superpower#155] [Q&A]
Greetings,

I have used Sprache prior for a small DSL (but non-trivial so Sprache helped alot with readability). At that point I was not aware of Superpower but now I have a choice as I embark on a quick implementation of the [minimal nestedtext](https://nestedtext.org/en/stable/minimal-nestedtext.html) format (which to my knowledge does not have a .net implementation).

In general, would you say I would gain anything from tokenizing with that syntax, but on the other hand, can I still do everything Sprache can with Superpower?

In specific, the nestedtext syntax is one of those that rely on the indent level and judging from a sibling discussion that is not straight forward to model with either of the monad based parsers. I can do this *relatively* straight forward in a PEG parser, such as my javascript implementation which utilize Peggy. Has there been any further thoughts on this since the time of that discussion?

**Comment by nblumhardt** [comment]
Hi! I haven't attempted an indentation-sensitive language parser with Superpower, so unfortunately I can't offer much advice, but at the outset I think I'd be looking to track indentation level in the tokenizer and to emit "fake" tokens for _increase indent_ and _decrease indent_, that would then play similar roles to `{` and `}` in the subsequent parser. No idea how much effort this would be relative to other approaches, though.

If you end up with something elegant, an example for the `examples/` folder would be more than welcome :-)

Cheers and good luck!

**Reply by RebelusQuo** [reply]
My immediate thought was something like that, to tokenize the indention. In my case I can't just ignore whitespace, but perhaps the built in tokenizer would benefit from an "ignore whitespace except at start of row" method? But that's a later adventure.

**Comment by RebelusQuo** [comment]
The bane of asking multiple questions in one; I also wanted your input on the compatibility between Sprache and Superpower. Would you say Superpower can do all that Sprache can, or do I have a choice to make here? What are the differences and pros/cons?

**Reply by nblumhardt** [reply]
Superpower has a text parser type and combinators that roughly match Sprache's - the intention is that Superpower is/will be roughly a superset, and it's the more active codebase. Some features might be missing, but they're generally easy to add atop what is there. HTH!

**Reply by RebelusQuo** [reply]
Then I'll make an attempt with Superpower and report back to you. Thanks for now!
